### PR TITLE
added `mwseMCMSetting:convertToLabelValue`

### PR DESCRIPTION
**Pull request by herbertgithubaccount**
this method is responsible for handling how variable values get converted to the values displayed in the MCM. some advantages of this are:
- trivializes making `Slider` subclasses
- allows the custom formatting options to be applied to labels that contain an `%s`. (previously this wasn't the case)
- makes it easy for users to specify their own slider conversion functions: they now get more control over how a config value translates to something shown in the MCM.
  - this would be done by passing a `convertToLabelValue = function(self, variableValue) ... end` parameter to the `create<Setting>` methods.
  - this makes it easy to do things like: 
    - make sliders that store distances using game units,
    - use a slider to store the values of `enum` tables, and then display the keys in the slider:
      ```lua 
      convertToLabelValue = function(self, value) return table.find(enumTable, value) end
      ```

### CODE REVIEW

File: misc/package/Data Files/MWSE/core/mcm/components/settings/YesNoButton.lua [hunk 1]
```diff
@@ -8,15 +8,16 @@
...
 --- @class mwseMCMYesNoButton
 local YesNoButton = Parent:new()
-YesNoButton.defaultSetting = false
 
-function YesNoButton:getText()
-	local text = (self.variable.value and tes3.findGMST(tes3.gmst.sYes).value or tes3.findGMST(tes3.gmst.sNo).value)
-	return text --[[@as string]]
+function YesNoButton:convertToLabelValue(variableValue)
+	return tes3.findGMST(variableValue and tes3.gmst.sYes or tes3.gmst.sNo).value
 end
```
Comment (C3pa) [inline misc/package/Data Files/MWSE/core/mcm/components/settings/YesNoButton.lua:21]: I've just found out that the base Component class stores these two GMST values as `sYes` and `sNo`, so could make it just:
`return variableValue and self.sYes or self.sNo`

File: misc/package/Data Files/MWSE/core/mcm/components/settings/OnOffButton.lua [hunk 2]
```diff
@@ -10,13 +10,13 @@
...
 --- @class mwseMCMOnOffButton
 local OnOffButton = Parent:new()
-OnOffButton.defaultSetting = false
 
-function OnOffButton:getText()
-	local text = (self.variable.value and tes3.findGMST(tes3.gmst.sOn).value or tes3.findGMST(tes3.gmst.sOff).value)
-	return text --[[@as string]]
+function OnOffButton:convertToLabelValue(variableValue)
+	return tes3.findGMST(variableValue and tes3.gmst.sOn or tes3.gmst.sOff).value
 end
```
Comment (C3pa) [inline misc/package/Data Files/MWSE/core/mcm/components/settings/OnOffButton.lua:20]: As I noted in YesNoButton, the base Component class stores `sYes` and `sNo`. Could you also add the `sOn` and `sOff` to the base Component class and use these values as fields here:
`return variableValue and self.sYes or self.sNo`

**Review by C3pa — COMMENTED** [review]
I like the proposal. I am not sure that DistanceSlider is general enough to be added as a standalone Setting, but it certainly is useful. I propose to add the implementation of DistanceSlider as an example to the docs for Slider. I think that way, if somebody needs exactly that kind of slider, they can take it straight from the docs, or even better, the example might help teach people to implement custom-behaving sliders.

**Comment by herbertgithubaccount** [comment]
> I like the proposal. I am not sure that DistanceSlider is general enough to be added as a standalone Setting, but it certainly is useful. I propose to add the implementation of DistanceSlider as an example to the docs for Slider. I think that way, if somebody needs exactly that kind of slider, they can take it straight from the docs, or even better, the example might help teach people to implement custom-behaving sliders.

thank you. i completely agree about the `DistanceSlider`. i'll take it out and add it as an example. that's a really good idea.

also, how do you feel about the name `convertToLabelValue`? do you think the name is fine or would something else be more appropriate if it's to be a parameter to the `create` methods? maybe something like `labelConverter`?

File: misc/package/Data Files/MWSE/core/mcm/components/settings/OnOffButton.lua (hunk 2 above)
Comment (herbertgithubaccount) [reply misc/package/Data Files/MWSE/core/mcm/components/settings/OnOffButton.lua:20]: oh that's really nice to know. i'll change both this and the `YesNoButton`

Comment (C3pa) [reply misc/package/Data Files/MWSE/core/mcm/components/settings/OnOffButton.lua]: Please add the `sYes` and `sNo` fields to the base Component class.

File: autocomplete/definitions/namedTypes/mwseMCMSlider/convertToLabelValue/SkillSlider.lua [hunk 3]
```diff
@@ -0,0 +1,14 @@
+local DistanceSlider = mwse.mcm.createSlider{
```
Comment (C3pa) [inline autocomplete/definitions/namedTypes/mwseMCMSlider/convertToLabelValue/SkillSlider.lua:1]: A nit: can you change the variable name to `SkillSlider`, or don't store the result of `mwse.mcm.createSlider` call?

File: autocomplete/definitions/namedTypes/mwseMCMSetting/convertToLabelValue.lua [hunk 4]
```diff
@@ -0,0 +1,7 @@
+return {
+	type = "method",
+	description = [[This function specifies how values stored in the `variable` field should correspond to values displayed by this setting.]],
+	arguments = {
+		{ name = "variableValue", type = "number" }
+	},
```
Comment (C3pa) [inline autocomplete/definitions/namedTypes/mwseMCMSetting/convertToLabelValue.lua:6]: This function has a return value.

File: autocomplete/definitions/namedTypes/mwseMCMSlider/convertToLabelValue.lua [hunk 5]
```diff
@@ -0,0 +1,21 @@
+return {
+	type = "method",
+	description = [[This function specifies how values stored in the `variable` field should correspond to values displayed in the slider label.
+The default behavior is to consistently format decimal places (i.e., if `decimalPlaces == 2`, make sure two decimal places are shown.)
+This can be overwritten in the `createNewSlider` method, allowing for custom formatting of variable values.]],
+	arguments = {
+		{ name = "variableValue", type = "number" }
+	},
+    examples = {
```
Comment (C3pa) [inline autocomplete/definitions/namedTypes/mwseMCMSlider/convertToLabelValue.lua:9]: This function has a return value.

**Comment by C3pa** [comment]
> also, how do you feel about the name convertToLabelValue? do you think the name is fine or would something else be more appropriate if it's to be a parameter to the create methods? maybe something like labelConverter?

I think `convertToLabelValue` is consistent with the naming of other methods on Slider. This method isn't only used in the Slider though, and I am also fine with `labelConverter`.

On a side note, maybe not made clear in our docs to the reader, but any method/field of an MCM component can be overridden on creation.

Also, the addition of examples is very welcome. We can insert the same example file into multiple places in the docs. In other definitions that would be:
```lua
    -- autocomplete\\definitions\\namedTypes\\tes3game\\clearTarget.lua
    examples = {
        -- Relative path to the example file
        [""..\\..\\..\\global\\tes3\\setTrap\\untrapping""] = { ... }
    }
```
I'd like to see these examples also linked at `mwse.mcm.createSlider` and `mwseMCMCategory:createSlider` pages. If you don't feel like it, I can do that after we merge this.

File: autocomplete/definitions/namedTypes/mwseMCMSlider/convertToLabelValue.lua (hunk 5 above)
Comment (herbertgithubaccount) [reply autocomplete/definitions/namedTypes/mwseMCMSlider/convertToLabelValue.lua:9]: looks like i got a little sloppy with the copy pasting last night. all three of these things got fixed. 

i went with your suggestion of not storing the results in the `convertToLabelValue` examples, i think that was a nice aesthetic choice.

**Comment by herbertgithubaccount** [comment]
i agree that it does synergize nicely with the other slider methods. i think my current preference is to keep it as `convertToLabelValue`. i thought about it a bit more and i think the name `labelConverter` would imply that it takes in one label value and spits out a different one, which might be misleading.

i also added the examples to both `createSlider` functions, that was a good idea. 

do you think i should add `convertToLabelValue` as an optional parameter to the `createSlider` methods?

**Comment by C3pa** [comment]
> do you think i should add convertToLabelValue as an optional parameter to the createSlider methods?

Sure, you can add it.

**Comment by herbertgithubaccount** [comment]
it has been done

**Review by C3pa — APPROVED** [review]

**Pull request merged**